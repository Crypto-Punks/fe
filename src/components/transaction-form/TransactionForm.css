.TransactionForm {
  background-color:  rgba(67, 72, 79, 0.897);
  border: none;
  display: flex;
  margin-bottom: 10%;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  height: 40%;
  width: 100%;
  z-index: 1;
  color: white;
  border-radius: .3em;

  label {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    margin: 10px;
  
    input {
      margin: 10px;
    }

    select {
      height: 2em;
      background-color: #BF60A5;
    }
  }

  section {
    text-align: center;
    width: 10vw;
    margin: .3em;
  }

  button {
    background-color: #BF60A5;
    border-radius: .3em;
  }
}

@media (max-width: 500px) {

  .TransactionForm {
    height: 50%;
    left: 0;
    width: 100vw;
    top: 4.461%;

    label {

      input {
        width: 40vw;
      }

      select {
        width: 45vw;
      }
    }

    section {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 100vw;

      input {
        margin: 10px;
        width: 40vw;
        height: 1.5em;
    }
  }
}
}
