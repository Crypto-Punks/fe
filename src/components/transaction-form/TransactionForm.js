import React, { useState, useEffect } from 'react';
import { evaluate } from 'mathjs';
import PropTypes from 'prop-types';
import { getCoinById } from '../../services/currencies';
import styles from './TransactionForm.css';


const TransactionForm = ({ handleSubmit, currencies, investedCoins }) => {
  const [fromCurrency, setFromCurrency] = useState('');
  const [fromCurrencyAmount, setFromCurrencyAmount] = useState(0);
  const [fromCurrencyMax, setFromCurrencyMax] = useState(0);
  
  const [toCurrency, setToCurrency] = useState('');
  const [toCurrencyAmount, setToCurrencyAmount] = useState(0);
  const [toCurrencyMax, setToCurrencyMax] = useState(0);
  
  const [exchangeRate, setExchangeRate] = useState(0);
  const [transactionValue, setTransactionValue] = useState(0);
  
  const currenciesElements = currencies.sort().map(currency => {
    return <option key={currency} value={currency}>{currency}</option>;
  });

  const investedCoinsElements = investedCoins.sort().map(coin => {
    if(coin.name !== 'USD') return <option key={coin.name} value={coin.name}>{coin.name}</option>;
  });

  useEffect(() => {
    if(fromCurrency) {
      const max = investedCoins.find(element => element.name === fromCurrency).amount;
      setFromCurrencyMax(max);
    }
  }, [fromCurrency]);

  useEffect(() => {
    if(fromCurrency && toCurrency) {
      Promise.all([
        getCoinById(fromCurrency),
        getCoinById(toCurrency)
      ])
        .then(([fromCurrencyResult, toCurrencyResult]) => {
          const exchangeRate = evaluate(Number(fromCurrencyResult[0].priceUsd) / Number(toCurrencyResult[0].priceUsd));
          setExchangeRate(exchangeRate);
          setFromCurrencyAmount(evaluate(toCurrencyAmount / exchangeRate));
        });
    }
  }, [fromCurrency]);

  useEffect(() => {
    if(fromCurrency && toCurrency) {
      Promise.all([
        getCoinById(fromCurrency),
        getCoinById(toCurrency)
      ])
        .then(([fromCurrencyResult, toCurrencyResult]) => {
          const exchangeRate = evaluate(Number(fromCurrencyResult[0].priceUsd) / Number(toCurrencyResult[0].priceUsd));
          setExchangeRate(exchangeRate);
          setToCurrencyAmount(evaluate(Number(fromCurrencyAmount * exchangeRate)));
        });
    }
  }, [toCurrency]);

  useEffect(() => {
    if(exchangeRate) {
      const max = evaluate(fromCurrencyMax * exchangeRate);
      setToCurrencyMax(max);
    }
  }, [toCurrencyAmount, fromCurrencyMax]);

  useEffect(() => {
    if(fromCurrency) {
      getCoinById(fromCurrency)
        .then(fromCurrencyResult => {
          setTransactionValue(evaluate(fromCurrencyResult[0].priceUsd * fromCurrencyAmount));
        });
    }
  }, [fromCurrency, fromCurrencyAmount]);

  useEffect(() => {
    if(toCurrency) {
      getCoinById(toCurrency)
        .then(toCurrencyResult => {
          setTransactionValue(evaluate(toCurrencyResult[0].priceUsd * toCurrencyAmount));
        });
    }
  }, [toCurrency, toCurrencyAmount]);

  return (
    <form className={styles.TransactionForm} onSubmit={event => handleSubmit(event, exchangeRate, toCurrency, toCurrencyAmount, fromCurrency, fromCurrencyAmount, investedCoins)}>
      <label>
        I want to buy
        <input 
          type='number' 
          value={toCurrencyAmount}  
          max={toCurrencyMax} 
          min={0}
          step={'any'} 
          onChange={event => {
            setToCurrencyAmount(event.target.value);
            if(exchangeRate) setFromCurrencyAmount(evaluate(event.target.value / exchangeRate));
          }} 
        />
        <select 
          value={toCurrency} 
          onChange={(event) => setToCurrency(event.target.value)}>
          <option value={''} disabled hidden>Choose your To Currency</option>
          <option key={'USD'} value={'USD'}>{'USD'}</option>
 
          {currenciesElements}
        </select>
      </label>
      <label>
        I want to use
        <input 
          type='number' 
          value={fromCurrencyAmount} 
          max={fromCurrencyMax} 
          min={0} 
          step={'any'} 
          onChange={event => {
            setFromCurrencyAmount(event.target.value);
            if(exchangeRate) setToCurrencyAmount(evaluate(event.target.value * exchangeRate));
          }} 
        />
        <select 
          value={fromCurrency} 
          onChange={event => setFromCurrency(event.target.value)}>
          <option value={''} disabled hidden>Choose Your From Currency</option>
          <option key={'USD'} value={'USD'}>{'USD'}</option>
          {investedCoinsElements}
        </select>
      </label>
      <section>
        <p>Exchange rate:</p>
        <input 
          type='number' 
          value={exchangeRate}
          readOnly={true} 
        />
      </section>
      <section>
        <p>Value of transaction (USD):</p>
        <input 
          type='number' 
          value={transactionValue} 
          readOnly={true}
        />
      </section>
      <button>Complete Your Transaction</button>
    </form>
  );
};

TransactionForm.propTypes = {
  investedCoins: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    amount: PropTypes.number,
  })).isRequired,
  currencies: PropTypes.arrayOf(PropTypes.string.isRequired),
  handleSubmit: PropTypes.func.isRequired
};


export default TransactionForm;
