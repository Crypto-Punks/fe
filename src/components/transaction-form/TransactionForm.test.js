import React from 'react';
import { shallow } from 'enzyme';
import TransactionForm from './TransactionForm';

const data = {
  handleSubmit: () => {},
  currencies: [],
  investedCoins: []
};

describe('TransactionForm component', () => {
  it('renders TransactionForm', () => {
    const wrapper = shallow(<TransactionForm { ...data} />);
    expect(wrapper).toMatchSnapshot();
  });
});
