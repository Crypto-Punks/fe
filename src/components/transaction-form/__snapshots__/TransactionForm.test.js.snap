// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionForm component renders TransactionForm 1`] = `
<form
  className="TransactionForm"
  onSubmit={[Function]}
>
  <label>
    <select
      onChange={[Function]}
      value=""
    >
      <option
        disabled={true}
        hidden={true}
        value=""
      >
        Choose Coin to Buy
      </option>
      <option
        key="USD"
        value="USD"
      >
        USD
      </option>
    </select>
    <input
      max={0}
      min={0}
      onChange={[Function]}
      placeholder="How much do you want to buy?"
      step="any"
      type="number"
      value={0}
    />
  </label>
  <label>
    <select
      onChange={[Function]}
      value=""
    >
      <option
        disabled={true}
        hidden={true}
        value=""
      >
        Paying With
      </option>
      <option
        key="USD"
        value="USD"
      >
        USD
      </option>
    </select>
    <input
      max={0}
      min={0}
      onChange={[Function]}
      placeholder="How much do you want to buy?"
      step="any"
      type="number"
      value={0}
    />
  </label>
  <section>
    <p>
      Exchange rate:
    </p>
    <input
      readOnly={true}
      type="number"
      value={0}
    />
  </section>
  <section>
    <p>
      Value (USD):
    </p>
    <input
      readOnly={true}
      type="number"
      value={0}
    />
  </section>
  <button>
    Complete Your Transaction
  </button>
</form>
`;
